### PR TITLE
Bug 1895360: pkg/daemon: don't delete a file if its replaced with a dropin

### PR DESCRIPTION
**Pull request by vrutkovs**
Some systemd service settings may be defined in .Storage.Files and in .Systemd.Units.Dropins. Dropins is preferable, so users may want to migrate to the way.

However if a file representing a dropin was defined in .Storage.Files and converted into .Systemd.Units.Dropin in new version of MC, it would be placed by Ignition's dropin implementation and then garbage-collected by MCD's deleteStaleData.

This change ensures we're checking Systemd.Units.Dropins before attempting to remove this file. This is crucial for OKD 4.5 -> 4.6 upgrades - kubelet MCO config was defined as a file in 4.5 and converted into a dropin in 4.6

### CODE REVIEW

**Comment by kikisdeliveryservice** [comment]
hey @vrutkovs this seems like it should have a test, can you add one?

**Comment by vrutkovs** [comment]
Sure, will add a unit test (verifying that it fixes OKD upgrade first). Not sure how to approach unittesting here - do you know any test which checks `deleteStaleFiles`?

**Comment by vrutkovs** [comment]
/cc @kikisdeliveryservice 

Added a unit test, verified that OKD 4.5 can be upgraded using MCO with this patch

**Comment by vrutkovs** [comment]
Flakes

/retest

**Comment by vrutkovs** [comment]
/skip

e2e-gcp-op is permafailing

File: pkg/daemon/update.go [hunk 1]
```diff
@@ -1073,55 +1106,63 @@ func (dn *Daemon) deleteStaleData(oldIgnConfig, newIgnConfig *ign3types.Config)
...
+			}
+
+			if err := os.Remove(origFileName(f.Path)); err != nil {
+				return errors.Wrapf(err, "deleting orig file %q: %v", origFileName(f.Path), err)
+			}
+		}
+
+		// Check Systemd.Units.Dropins - don't remove the file if configuration has been converted into a dropin
+		if dn.isPathInDropins(f.Path, &newIgnConfig.Systemd) {
+			glog.V(2).Infof("Not removing file %q: replaced with systemd dropin", f.Path)
```
Comment (yuqi-zhang) [inline pkg/daemon/update.go:1152]: Probably should just log with default verbosity here. Shouldn't happen very often and would be nice for us to know if this causes some edge case in the future

**Review by yuqi-zhang — COMMENTED** [review]
I think this patch should fix the bug in question, however this doesn't seem very bulletproof. Consider the following scenarios:

1. systemd units in the unit section would have this problem as well right? if a systemd unit that was originally being written as a file gets replaced by a unit section, we should also correctly handle that.
2. the opposite case is ironically not handled as well, where a dropin that was replaced by a file would also not work (why would you ever do this though). I guess in general anything cross-dependency between unit and file sections is not handled well.
3. if there was a dropin originally on the system (call it version A). which the MCD made a backup of when you wrote the file via a MC (call it version B), and then you try to replace with a dropin (call it version C), it looks like the logic will end up writing version A to the file instead of version C like we want, since we perform a backup and then a no-op. A rare edge case but maybe we should account for it anyways for correctness.

Those 3 scenarios are just some that came to mind (my assessment might not be entirely correct here). I am not opposed to merging this fix as is but we should consider the interactions between files and unit sections a bit deeper if that's the route we take. I'd be ok to merge this as is as well (maybe fix scenario 3?) provided we don't regress any behaviour

File: pkg/daemon/update.go [hunk 2]
```diff
@@ -1073,55 +1106,63 @@ func (dn *Daemon) deleteStaleData(oldIgnConfig, newIgnConfig *ign3types.Config)
...
+				continue
+			}
+
+			if err := os.Remove(origFileName(f.Path)); err != nil {
+				return errors.Wrapf(err, "deleting orig file %q: %v", origFileName(f.Path), err)
+			}
+		}
+
+		// Check Systemd.Units.Dropins - don't remove the file if configuration has been converted into a dropin
+		if dn.isPathInDropins(f.Path, &newIgnConfig.Systemd) {
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply pkg/daemon/update.go:1151]: Right. I had to rework the whole function as `gosec` was complaining about nested `if`s in [`e79a186` (#2196)](https://github.com/openshift/machine-config-operator/pull/2196/commits/e79a186a4ecf039ffa33b03cc7187bac745d363a)

**Comment by vrutkovs** [comment]
> this doesn't seem very bulletproof.

Sadly there are lots of edge cases, lets focus on this one as it blocked 4.5 -> 4.6 OKD upgrade and thus tagging OKD 4.6 in stable for now.

>systemd units in the unit section would have this problem as well right?

Correct, however simply landing `.service` won't turn it into a full-blown service, would it? It would happen only after the reboot.

>a dropin that was replaced by a file would also not work

Correct, replacing a dropin with a file would need a fix further in the function. I agree that this might be the rarest case

>it looks like the logic will end up writing version A to the file instead of version C like we want

This change doesn't change the contents logic - it would be written by Ignition: rewrite the file with version C and not remove the file as its no longer in `.storage.files`, but defined in `systemd.units` now

File: pkg/daemon/update.go (hunk 1 above)
Comment (vrutkovs) [reply pkg/daemon/update.go]: Agree, using the same verbosity as in `Removed stale file` message

File: pkg/daemon/update.go [hunk 3]
```diff
@@ -1055,6 +1055,39 @@ func restorePath(path string) error {
...
+// Returns is dropin (true/false), service name, dropin name
+func isPathASystemdDropin(path string) (bool, string, string) {
+	if !strings.HasPrefix(path, "/etc/systemd/system") {
+		return false, "", ""
+	}
+	pathSegments := strings.Split(path, "/")
+	dropinName := pathSegments[len(pathSegments)-1]
+	servicePart := pathSegments[len(pathSegments)-2]
+	allServiceSegments := strings.Split(servicePart, ".")
+	if allServiceSegments[len(allServiceSegments)-1] != "d" {
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply pkg/daemon/update.go:1071]: Fixed by adding `HasSuffix` check

**Comment by yuqi-zhang** [comment]
> This change doesn't change the contents logic - it would be written by Ignition: rewrite the file with version C and not remove the file as its no longer in .storage.files, but defined in systemd.units now

So in the scenario I described, we would be shipping a dropin on the base OS, not in the installed ignition configs. Then when you defined your own unit, it saved what was originally on system as a .orig file. Now when you change it, what would happen is that when we hit these conditions:

```
if _, ok := newFileSet[f.Path]; ok {                                                     <-  we do not see it in the new file, skips this

if _, err := os.Stat(noOrigFileStampName(f.Path)); err == nil {         <- there is an orig, skip this

else if _, err := os.Stat(origFileName(f.Path)); err == nil {                 <- we hit this condition and restore the old file
```

Then when it hits `isPathInDropins` its also a true, so we do not do anything. Since this happened after the unit write happens, we restore the original, not the newest one you wrote.

I'm probably being overly cautious and this won't happen today.

**Review by yuqi-zhang — APPROVED** [review]
Approving the functionality. Will give a chance for others to take a look

File: pkg/daemon/update.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sinnykumari) [reply pkg/daemon/update.go:1154]: the way github is showing diff is indeed confusing

**Comment by sinnykumari** [comment]
overall logic looks fine to me, hopefully we are not skipping any corner case.
/lgtm

**Comment by vrutkovs** [comment]
/hold

Not sure why gcp-op won't pass. Holding it for >4 hours, so that fresh release image would be built

**Comment by kikisdeliveryservice** [comment]
> /hold
> 
> Not sure why gcp-op won't pass. Holding it for >4 hours, so that fresh release image would be built

we are waiting for https://github.com/openshift/machine-config-operator/pull/2229 to merge so feel free to hold it and check it tomorrow or later so avoid endless retests

**Pull request merged**